drivers/ethernet/eth_numaker.c

```diff
@@ -0,0 +1,807 @@
+		gmacdev->DuplexMode = FULLDUPLEX;
+		gmacdev->Speed = SPEED100;
+		synopGMAC_set_full_duplex(gmacdev);
+		synopGMAC_set_mode(NUMAKER_GMAC_INTF, 1); /* Set mode 1: 100Mbps; 2: 10Mbps */
+	} else if (reg & ADVERTISE_100HALF) {
+		LOG_DBG("100 half");
+		gmacdev->DuplexMode = HALFDUPLEX;
+		gmacdev->Speed = SPEED100;
+		synopGMAC_set_half_duplex(gmacdev);
+		synopGMAC_set_mode(NUMAKER_GMAC_INTF, 1); /* Set mode 1: 100Mbps; 2: 10Mbps */
+	} else if (reg & ADVERTISE_10FULL) {
+		LOG_DBG("10 full");
+		gmacdev->DuplexMode = FULLDUPLEX;
+		gmacdev->Speed = SPEED10;
+		synopGMAC_set_full_duplex(gmacdev);
+		synopGMAC_set_mode(NUMAKER_GMAC_INTF, 2); /* Set mode 1: 100Mbps; 2: 10Mbps */
+	} else {
+		LOG_DBG("10 half");
+		gmacdev->DuplexMode = HALFDUPLEX;
+		gmacdev->Speed = SPEED10;
+		synopGMAC_set_half_duplex(gmacdev);
+		synopGMAC_set_mode(NUMAKER_GMAC_INTF, 2); /* Set mode 1: 100Mbps; 2: 10Mbps */
+	}
+
+	LOG_INF("PHY ID 1:0x%x", mdio_read(gmacdev, eth_phy_addr, MII_PHYSID1));
+	LOG_INF("PHY ID 2:0x%x", mdio_read(gmacdev, eth_phy_addr, MII_PHYSID2));
+
+	return status;
+}
+
+static void m_numaker_read_mac_addr(char *mac)
+{
+	uint32_t uid1;
+	/* Fetch word 0 of data flash */
+	uint32_t word0 = *(uint32_t *)(NUMAKER_DATA_FLASH + 0x04U);
+	/*
+	 * Fetch word 1 of data flash
+	 * we only want bottom 16 bits of word1 (MAC bits 32-47)
+	 * and bit 9 forced to 1, bit 8 forced to 0
+	 * Locally administered MAC, reduced conflicts
+	 * http://en.wikipedia.org/wiki/MAC_address
+	 */
+	uint32_t word1 = *(uint32_t *)NUMAKER_DATA_FLASH;
+
+	/* Not burn any mac address at the beginning of data flash */
+	if (word0 == NUMAKER_MASK_32) {
+		/* Generate a semi-unique MAC address from the UUID */
+		SYS_UnlockReg();
+		/* Enable FMC ISP function */
+		FMC_Open();
+		uid1 = FMC_ReadUID(1);
+		word1 = (uid1 & 0x003FFFFF) | ((uid1 & 0x030000) << 6) >> 8;
+		word0 = ((FMC_ReadUID(0) >> 4) << 20) | ((uid1 & 0xFF) << 12) |
+			(FMC_ReadUID(2) & 0xFFF);
+		/* Disable FMC ISP function */
+		FMC_Close();
+		/* Lock protected registers */
+		SYS_LockReg();
+	}
+
+	word1 |= 0x00000200;
+	word1 &= 0x0000FEFF;
+
+	mac[0] = (word1 & 0x0000ff00) >> 8;
+	mac[1] = (word1 & 0x000000ff);
+	mac[2] = (word0 & 0xff000000) >> 24;
+	mac[3] = (word0 & 0x00ff0000) >> 16;
+	mac[4] = (word0 & 0x0000ff00) >> 8;
+	mac[5] = (word0 & 0x000000ff);
+
+	LOG_INF("mac address %02x-%02x-%02x-%02x-%02x-%02x", mac[0], mac[1], mac[2], mac[3],
+	       mac[4], mac[5]);
+}
+
+static void m_numaker_gmacdev_enable(synopGMACdevice *gmacdev)
+{
+
+	synopGMAC_clear_interrupt(gmacdev);
+
+	/* Enable INT & TX/RX */
+	synopGMAC_enable_interrupt(gmacdev, DmaIntEnable);
+	synopGMAC_enable_dma_rx(gmacdev);
+	synopGMAC_enable_dma_tx(gmacdev);
+
+	synopGMAC_tx_enable(gmacdev);
+	synopGMAC_rx_enable(gmacdev);
+}
+
+static int m_numaker_gmacdev_init(synopGMACdevice *gmacdev, uint8_t *mac_addr, uint32_t gmac_base)
+{
+	int status;
+	int i;
+	uint32_t offload_needed = 0;
+	struct sk_buff *skb;
+
+	LOG_DBG("---Start---");
```

Comment:
nit: drop the `---`, it's just a debug print, no point doing too much ascii graphics around it, you may even just do `LOG_DBG("");` since the function name is printed in debug prints, same information, no wasted flash.